core/bessctl.cc

```diff
@@ -1640,38 +1634,58 @@ class BESSControlImpl final : public BESSControl::Service {
 
     return c;
   }
-};
 
-// TODO: C++-ify
-static std::unique_ptr<Server> server;
-static BESSControlImpl service;
+  // function to call to close this gRPC service.
+  std::function<void()> shutdown_func_;
+};
 
-void SetupControl() {
-  ServerBuilder builder;
+bool ApiServer::grpc_cb_set_ = false;
 
-  if (FLAGS_p) {
-    std::string server_address = bess::utils::Format("127.0.0.1:%d", FLAGS_p);
-    LOG(INFO) << "Server listening on " << server_address;
-    builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
+void ApiServer::Listen(const std::string &host, int port) {
+  if (!builder_) {
+    builder_ = new grpc::ServerBuilder();
   }
 
-  builder.RegisterService(&service);
-  server = builder.BuildAndStart();
+  std::string addr = bess::utils::Format("%s:%d", host.c_str(), port);
+  LOG(INFO) << "Server listening on " << addr;
+
+  builder_->AddListeningPort(addr, grpc::InsecureServerCredentials());
 }
 
-void RunControl() {
-  auto serve_func = []() {
-    server->Wait();
-    LOG(INFO) << "Terminating gRPC server";
+void ApiServer::Run() {
+  class ServerCallbacks : public grpc::Server::GlobalCallbacks {
+   public:
+    ServerCallbacks() {}
+    void PreSynchronousRequest(ServerContext*) { mutex_.lock(); }
+    void PostSynchronousRequest(ServerContext*) { mutex_.unlock(); }
```

Comment:
Any hard crash in a gRPC handler thread will terminate the entire daemon process. Otherwise, the callback function will be called after any soft failures (e.g., not returning Status::OK).